ouroboros-network/src/Ouroboros/Network/TxSubmission/Outbound.hs

```diff
@@ -110,60 +110,60 @@ txSubmissionOutbound tracer maxUnacked TxSubmissionMempoolReader{..} _version co
                 do
                   MempoolSnapshot{mempoolTxIdsAfter} <- mempoolGetSnapshot
                   let txs = mempoolTxIdsAfter lastIdx
                   check (not $ null txs)
                   pure (take (fromIntegral reqNo) txs)
 
+              case mbtxs of
+                Nothing -> pure (SendMsgDone ())
+                Just txs ->
+                  let !(txs', client') = update txs
+                      txs'' = case NonEmpty.nonEmpty txs' of
+                        Just x -> x
+                        -- Assert txs is non-empty: we blocked until txs was non-null,
+                        -- and we know reqNo > 0, hence `take reqNo txs` is non-null.
+                        Nothing -> error "txSubmissionOutbound: empty transaction's list"
+                  in  pure (SendMsgReplyTxIds (BlockingReply txs'') client')
+
             TokNonBlocking -> do
               when (reqNo == 0 && ackNo == 0) $
                 throwIO ProtocolErrorRequestedNothing
               when (Seq.null unackedSeq') $
                 throwIO ProtocolErrorRequestNonBlocking
 
-              atomically $ do
+              txs <- atomically $ do
                 MempoolSnapshot{mempoolTxIdsAfter} <- mempoolGetSnapshot
                 let txs = mempoolTxIdsAfter lastIdx
-                return (Just $ take (fromIntegral reqNo) txs)
-
-          return $! case (mbtxs, blocking) of
-            (Nothing, TokBlocking)    -> SendMsgDone ()
-            (Nothing, TokNonBlocking) -> error "txSubmissionOutbound: impossible happend!"
-            (Just txs, _) ->
-              -- These txs should all be fresh
-              assert (all (\(_, idx, _) -> idx > lastIdx) txs) $
-                -- Update our tracking state with any extra txs available.
-                let !unackedSeq'' = unackedSeq' <> Seq.fromList
-                                      [ (txid, idx) | (txid, idx, _) <- txs ]
-                    !lastIdx'
-                      | null txs  = lastIdx
-                      | otherwise = idx where (_, idx, _) = last txs
-                    txs'         :: [(txid, TxSizeInBytes)]
-                    txs'          = [ (txid, size) | (txid, _, size) <- txs ]
-                    client'       = client unackedSeq'' lastIdx'
-
-                -- Our reply type is different in the blocking vs non-blocking cases
-                in case blocking of
-                    TokNonBlocking -> SendMsgReplyTxIds (NonBlockingReply txs') client'
-                    TokBlocking    -> SendMsgReplyTxIds (BlockingReply   txs'') client'
-                      where
-                        txs'' = case NonEmpty.nonEmpty txs' of
-                          Just  x -> x
-                          Nothing -> error "txSubmissionOutbound: empty transaction's list"
-                        -- Assert txs is non-empty: we blocked until txs was non-null,
-                        -- and we know reqNo > 0, hence take reqNo txs is non-null.
```

Comment:
I'd appreciate a second look at my refactor here, in particular w.r.t. whether or not I preserved the existing strictness properties.

Reply:
Strictness looks fine, but I am not sure if the logic is preserved. Previously we scrutinised `mbtxs` in `TokBlocking` case and in `TokNonBlocking` (we only errored if it was `Nothing`); now we don't scrutinise it the `ToknonBlocking` one at all.

Reply:
In the unmodified code, there's no path that allows constructing `mbtxs` via `Nothing` if `blocking` is `TokNonBlocking`, so I don't think it actually needs to be scrutinized. That's the crux of my decision to refactor. Does that strike you as valid?

Reply:
You're right, it looks good.